File: lua/glean/lineage.lua
-- glean.lineage: compose an ordered sequence of commit patches into per-line
-- ownership for the combined (base..target) view.
--
-- Pure: no git, no nvim API. The input is the ordered list of patches produced
-- by a first-parent `git log -p -U0 -M` walk (parsed by glean.diff); the output
-- is, per path, the two maps the review model addresses lines by:
--   prov[new_lnum]     = { sha, lnum }  -- who wrote this surviving line, and
--                                          its number in that commit's blob
--   del_attr[old_lnum] = { sha, lnum }  -- who removed this base line, and its
--                                          number in that commit's pre-image
--
-- The walk models each path as an ordered list of *segments*, each covering a
-- run of consecutive lines of the file as it stands after the patches applied
-- so far:
--   { base = B, n = k }            k lines inherited from the base image at B
--   { sha = C, lnum = L, n = k }   k lines added by C at L in C's post-image
-- The list ends with an open-ended base segment (n = nil, "and the rest of the
-- file") so the file's length is never needed. One hunk is one splice, so the
-- structure is O(hunks) rather than O(file size).
local M = {}

local INF = math.huge

local function origin_at(s, off)
  if s.base then return { base = s.base + off } end
  return { sha = s.sha, lnum = s.lnum + off }
end

-- A sub-segment of `s` starting `off` lines in, covering `count` lines
-- (count == nil means open-ended).
local function slice(s, off, count)
  local out = origin_at(s, off)
  out.n = count
  return out
end

-- The initial state of a path that exists in the base image: every line is a
-- base line, at the same number.
function M.base_state()
  return { { base = 1 } }
end

-- Cut `remove_n` lines starting at position `at` out of `segments` and
-- substitute `seg` (which may be nil or zero-width for a pure deletion).
-- Returns the new segment list and the ordered list of displaced origins.
function M.splice(segments, at, remove_n, seg)
  local res, displaced = {}, {}
  local cut_end = at + remove_n
  local inserted = false

  local function push(s)
    if s and (s.n == nil or s.n > 0) then res[#res + 1] = s end
  end
  local function insert_seg()
    if inserted then return end
    inserted = true
    if seg and (seg.n or 0) > 0 then res[#res + 1] = seg end
  end

  local pos = 1
  for _, s in ipairs(segments) do
    local inf = s.n == nil
    local n = inf and INF or s.n
    local s_end = pos + n

    local head_n = math.min(n, math.max(0, at - pos))
    if head_n > 0 then push(slice(s, 0, head_n)) end

    if at <= s_end then insert_seg() end

    local mid_start = math.max(pos, at)
    local mid_end = math.min(s_end, cut_end)
    for k = mid_start, mid_end - 1 do
      displaced[#displaced + 1] = origin_at(s, k - pos)
    end

    local tail_start = math.max(pos, cut_end)
    if tail_start < s_end then
      -- `inf and nil or x` would always take the `or` branch (nil is falsy),
      -- turning the open-ended tail into a segment of length inf.
      local tail_n = nil
      if not inf then tail_n = s_end - tail_start end
      push(slice(s, tail_start - pos, tail_n))
    end

    if inf then break end
    pos = s_end
  end
  insert_seg()

  return res, displaced
end

-- Debug/testing helper: the origins of lines 1..n as a plain array.
function M.expand(segments, n)
  local out = {}
  local pos = 1
  for _, s in ipairs(segments) do
    local count = s.n or (n - pos + 1)
    for i = 1, count do
      if pos > n then return out end
      out[pos] = origin_at(s, i - 1)
      pos = pos + 1
    end
  end
  return out
end

-- Split a hunk's lines into contiguous change blocks (with -U0 there is exactly
-- one, but context lines are tolerated). Each block is
--   { at = <pre-image position>, dels = k, lnum = <new_lnum of first add>, adds = m }
local function blocks_of(hunk)
  local out = {}
  -- For a pure insertion git reports old_start as the line the text goes
  -- *after*, so the zero-width insert lands at old_start + 1.
  local next_old = hunk.old_count == 0 and hunk.old_start + 1 or hunk.old_start
  local cur = nil
  local function flush()
    if cur then out[#out + 1] = cur end
    cur = nil
  end
  for _, dl in ipairs(hunk.lines) do
    if dl.kind == "context" then
      flush()
      next_old = dl.old_lnum + 1
    elseif dl.kind == "del" then
      if cur and cur.adds > 0 then flush() end
      cur = cur or { at = dl.old_lnum, dels = 0, adds = 0 }
      cur.dels = cur.dels + 1
      next_old = dl.old_lnum + 1
    elseif dl.kind == "add" then
      cur = cur or { at = next_old, dels = 0, adds = 0 }
      cur.lnum = cur.lnum or dl.new_lnum
      cur.adds = cur.adds + 1
    end
  end
  flush()
  return out
end

-- Apply one commit's FileEntry to one path's state.
--   state = { segs = <segment list>, del_attr = <map> }
-- Hunk positions are expressed against the commit's *pre-image*, so a running
-- offset translates them into the (already partially spliced) running image.
function M.apply(state, sha, file)
  local offset = 0
  for _, hunk in ipairs(file.hunks or {}) do
    for _, b in ipairs(blocks_of(hunk)) do
      local seg = b.adds > 0 and { sha = sha, lnum = b.lnum, n = b.adds } or nil
      local segs, displaced = M.splice(state.segs, b.at + offset, b.dels, seg)
      state.segs = segs
      for i, origin in ipairs(displaced) do
        if origin.base then
          state.del_attr[origin.base] = { sha = sha, lnum = b.at + i - 1 }
        end
      end
      offset = offset + b.adds - b.dels
    end
  end
  return state
end

-- Apply an ordered (oldest-first) list of patches `{ sha, files }` on top of
-- `states` (the per-path segment/attribution state produced by earlier
-- patches), mutating and returning it. Splitting this out of `compose` lets a
-- refresh keep the committed layers' state and re-apply only the work-tree
-- layer over a clone of it.
function M.extend(states, patches)
  local function state_for(path, kind)
    local st = states[path]
    if not st then
      st = { segs = kind == "add" and {} or M.base_state(), del_attr = {} }
      states[path] = st
    end
    return st
  end

  for _, patch in ipairs(patches) do
    for _, file in ipairs(patch.files or {}) do
      if file.old_path and file.old_path ~= file.path and file.old_path ~= "/dev/null" then
        local prev = states[file.old_path]
        if prev then
          states[file.path] = prev
          states[file.old_path] = nil
        end
      end
      local st = state_for(file.path, file.kind)
      M.apply(st, patch.sha, file)
      if file.kind == "delete" then
        -- A re-add must start from nothing: no origin survives the delete, and
        -- the base lines have just been attributed to the deleter.
        states[file.path] = { segs = {}, del_attr = st.del_attr }
      end
    end
  end

  return states
end

-- Deep-copy a states table, so a layer can be applied on top without
-- destroying the state the copy was taken from.
function M.clone(states)
  local out = {}
  for path, st in pairs(states) do
    local segs, del_attr = {}, {}
    for i, s in ipairs(st.segs) do segs[i] = { base = s.base, sha = s.sha, lnum = s.lnum, n = s.n } end
    for k, v in pairs(st.del_attr) do del_attr[k] = v end
    out[path] = { segs = segs, del_attr = del_attr }
  end
  return out
end

-- Project composed states into the two maps the review model addresses lines
-- by: { [path] = { prov = {...}, del_attr = {...} } }.
function M.finish(states)
  local out = {}
  for path, st in pairs(states) do
    local prov = {}
    local pos = 1
    for _, s in ipairs(st.segs) do
      if s.n == nil then break end
      if not s.base then
        for i = 1, s.n do prov[pos + i - 1] = { sha = s.sha, lnum = s.lnum + i - 1 } end
      end
      pos = pos + s.n
    end
    out[path] = { prov = prov, del_attr = st.del_attr }
  end
  return out
end

-- Compose an ordered (oldest-first) list of patches into the per-path maps.
function M.compose(patches)
  return M.finish(M.extend({}, patches))
end

return M

File: node/core/lineage.ts
/**
 * Compose an ordered sequence of commit patches into per-line ownership for
 * the combined (base..target) view. Pure: the input is a first-parent
 * `git log -p -U0 -M` walk parsed by `diff.ts`.
 *
 * Each path is an ordered list of segments covering consecutive lines of the
 * file as it stands after the patches applied so far. The list ends with an
 * open-ended base segment, so the file's length is never needed and the
 * structure stays O(hunks) rather than O(lines).
 */
import type { FileEntry, Hunk } from "./diff.ts";
import type { Layer, PostLnum, PreLnum, RepoPath } from "./types.ts";

export type Origin =
  | { kind: "base"; base: number }
  | { kind: "commit"; sha: Layer; lnum: number };
/** `"open"` means "and the rest of the file". */
export type SegLen = number | "open";
export type Segment = Origin & { n: SegLen };

export type Attribution<L extends number> = { sha: Layer; lnum: L };
export type PathState = {
  segs: Segment[];
  /** base lnum → who removed it, and its number in that commit's pre-image. */
  delAttr: Map<number, Attribution<PreLnum>>;
};
export type States = Map<RepoPath, PathState>;
export type Patch = { sha: Layer; files: readonly FileEntry[] };

function originAt(s: Origin, off: number): Origin {
  return s.kind === "base"
    ? { kind: "base", base: s.base + off }
    : { kind: "commit", sha: s.sha, lnum: s.lnum + off };
}

function slice(s: Origin, off: number, n: SegLen): Segment {
  return { ...originAt(s, off), n };
}

const lenOf = (n: SegLen) => (n === "open" ? Infinity : n);

export function baseState(): Segment[] {
  return [{ kind: "base", base: 1, n: "open" }];
}

/**
 * Cut `removeN` lines at position `at` (1-based) out of `segments` and
 * substitute `seg`. Returns the new list and the displaced origins, in order.
 */
export function splice(
  segments: readonly Segment[],
  at: number,
  removeN: number,
  seg: Segment | undefined,
): { segs: Segment[]; displaced: Origin[] } {
  const res: Segment[] = [];
  const displaced: Origin[] = [];
  const cutEnd = at + removeN;
  let inserted = false;
  const push = (s: Segment) => {
    if (s.n === "open" || s.n > 0) res.push(s);
  };
  const insertSeg = () => {
    if (inserted) return;
    inserted = true;
    if (seg && lenOf(seg.n) > 0) res.push(seg);
  };

  let pos = 1;
  for (const s of segments) {
    const n = lenOf(s.n);
    const sEnd = pos + n;
    const headN = Math.min(n, Math.max(0, at - pos));
    if (headN > 0) push(slice(s, 0, headN));
    if (at <= sEnd) insertSeg();
    const midEnd = Math.min(sEnd, cutEnd);
    for (let k = Math.max(pos, at); k < midEnd; k++) {
      displaced.push(originAt(s, k - pos));
    }
    const tailStart = Math.max(pos, cutEnd);
    if (tailStart < sEnd) {
      push(
        slice(s, tailStart - pos, s.n === "open" ? "open" : sEnd - tailStart),
      );
    }
    if (s.n === "open") break;
    pos = sEnd;
  }
  insertSeg();
  return { segs: res, displaced };
}

/** Testing helper: origins of lines 1..n. */
export function expand(segments: readonly Segment[], n: number): Origin[] {
  const out: Origin[] = [];
  for (const s of segments) {
    const count = s.n === "open" ? n - out.length : s.n;
    for (let i = 0; i < count; i++) {
      if (out.length >= n) return out;
      out.push(originAt(s, i));
    }
  }
  return out;
}

type Block = { at: number; dels: number; adds: number; lnum: number };

/** A hunk's contiguous change blocks (one under -U0; context is tolerated). */
function blocksOf(hunk: Hunk): Block[] {
  const out: Block[] = [];
  // For a pure insertion git reports oldStart as the line the text goes
  // *after*, so the zero-width insert lands one later.
  let nextOld = hunk.oldCount === 0 ? hunk.oldStart + 1 : hunk.oldStart;
  let cur: Block | undefined;
  const flush = () => {
    if (cur) out.push(cur);
    cur = undefined;
  };
  for (const dl of hunk.lines) {
    if (dl.kind === "context") {
      flush();
      nextOld = dl.oldLnum + 1;
    } else if (dl.kind === "del") {
      if (cur && cur.adds > 0) flush();
      cur ??= { at: dl.oldLnum, dels: 0, adds: 0, lnum: 0 };
      cur.dels++;
      nextOld = dl.oldLnum + 1;
    } else {
      cur ??= { at: nextOld, dels: 0, adds: 0, lnum: 0 };
      if (cur.adds === 0) cur.lnum = dl.newLnum;
      cur.adds++;
    }
  }
  flush();
  return out;
}

/**
 * Apply one commit's FileEntry to one path's state. Hunk positions are in the
 * commit's pre-image, so a running offset maps them into the running image.
 */
export function apply(
  state: PathState,
  sha: Layer,
  file: FileEntry,
): PathState {
  let offset = 0;
  for (const hunk of file.hunks) {
    for (const b of blocksOf(hunk)) {
      const seg: Segment | undefined =
        b.adds > 0
          ? { kind: "commit", sha, lnum: b.lnum, n: b.adds }
          : undefined;
      const { segs, displaced } = splice(
        state.segs,
        b.at + offset,
        b.dels,
        seg,
      );
      state.segs = segs;
      displaced.forEach((origin, i) => {
        if (origin.kind === "base") {
          state.delAttr.set(origin.base, { sha, lnum: (b.at + i) as PreLnum });
        }
      });
      offset += b.adds - b.dels;
    }
  }
  return state;
}

/** Apply oldest-first `patches` on top of `states`, mutating and returning it. */
export function extend(states: States, patches: readonly Patch[]): States {
  for (const patch of patches) {
    for (const file of patch.files) {
      if (file.oldPath !== file.path && file.oldPath !== "/dev/null") {
        const prev = states.get(file.oldPath);
        if (prev) {
          states.set(file.path, prev);
          states.delete(file.oldPath);
        }
      }
      let st = states.get(file.path);
      if (!st) {
        st = {
          segs: file.kind === "add" ? [] : baseState(),
          delAttr: new Map(),
        };
        states.set(file.path, st);
      }
      apply(st, patch.sha, file);
      if (file.kind === "delete") {
        // A re-add starts from nothing: no origin survives the delete.
        states.set(file.path, { segs: [], delAttr: st.delAttr });
      }
    }
  }
  return states;
}

export function clone(states: States): States {
  const out: States = new Map();
  for (const [path, st] of states) {
    out.set(path, {
      segs: st.segs.map((s) => ({ ...s })),
      delAttr: new Map(st.delAttr),
    });
  }
  return out;
}

export type PathLineage = {
  /** Target-image line → the commit that wrote it, in its post-image. */
  prov: Map<number, Attribution<PostLnum>>;
  /**
   * Target-image line → the base-image line it was inherited from. A lazy
   * O(segments) lookup: the open tail must never be materialized.
   */
  base: (lnum: number) => number | undefined;
  delAttr: Map<number, Attribution<PreLnum>>;
};

function baseView(segs: readonly Segment[]): PathLineage["base"] {
  const ranges: { pos: number; n: SegLen; base: number }[] = [];
  let pos = 1;
  for (const s of segs) {
    if (s.kind === "base") ranges.push({ pos, n: s.n, base: s.base });
    if (s.n === "open") break;
    pos += s.n;
  }
  return (lnum) => {
    for (const r of ranges) {
      if (lnum >= r.pos && lnum < r.pos + lenOf(r.n))
        return r.base + lnum - r.pos;
    }
    return undefined;
  };
}

export function finish(states: States): Map<RepoPath, PathLineage> {
  const out = new Map<RepoPath, PathLineage>();
  for (const [path, st] of states) {
    const prov = new Map<number, Attribution<PostLnum>>();
    let pos = 1;
    for (const s of st.segs) {
      if (s.n === "open") break;
      if (s.kind === "commit") {
        for (let i = 0; i < s.n; i++) {
          prov.set(pos + i, { sha: s.sha, lnum: (s.lnum + i) as PostLnum });
        }
      }
      pos += s.n;
    }
    out.set(path, { prov, base: baseView(st.segs), delAttr: st.delAttr });
  }
  return out;
}

export function compose(patches: readonly Patch[]): Map<RepoPath, PathLineage> {
  return finish(extend(new Map(), patches));
}
